skimage/_shared/utils.py

```diff
@@ -726,9 +726,12 @@ def _supported_float_type(input_dtype, allow_complex=False):
     float_type : dtype
         Floating-point dtype for the image.
     """
+    try:
+        input_dtype = np.dtype(input_dtype)
+    except (TypeError, ValueError):
+        input_dtype = getattr(input_dtype, 'dtype', input_dtype)
     if isinstance(input_dtype, Iterable) and not isinstance(input_dtype, str):
         return np.result_type(*(_supported_float_type(d) for d in input_dtype))
```

Comment:
Oh, seems indeed a bit more complicated. Thanks for iterating on this. While this seems to work I'm not happy with the logic flow here as it's starting to be hard to reason about it. E.g. I think `isinstance(input_dtype, str)` is obsolete with the current solution but I'm not sure if there is truly no case where it is reached.

Rather than trying to handle all kinds of inputs we should probably be more restrictive. How about
```python
    if isinstance(input_dtype, tuple):
        return np.result_type(*(_supported_float_type(d) for d in input_dtype))
    input_dtype = np.dtype(input_dtype)
```
which just makes sure to fail explicitly for everything except dtypes or tuple or dtypes. That avoids iteration on arrays and also makes testing for `str` unnecessary. This would probably require updating the docstring a bit and some tests. Do you agree?

Reply:
Done, this is much cleaner indeed. 
With this, I also found that the `restoration.uft.ir2tf` function also called `_supported_float_type(imp_resp)` instead of `_supported_float_type(imp_resp.dtype)`, so I've updated this function as well.

Reply:
Perfect! Could you update the docstring as well? :pray: Except for that looks ready to merge.

Reply:
done :slightly_smiling_face: